tests/event.test.ts

```diff
@@ -0,0 +1,225 @@
+import { validate } from 'class-validator';
+import { plainToClass } from 'class-transformer';
+import { DatabaseConnection, EventFactory, FeedbackFactory, FileFactory, PortalState, UserFactory } from './data';
+import { ControllerFactory } from './controllers';
+import { UserAccessType } from '../types';
+import { Config } from '../config';
+import { StorageUtils } from './utils';
+import { SubmitEventFeedbackRequest } from '../api/validators/EventControllerRequests';
+
+beforeAll(async () => {
+  await DatabaseConnection.connect();
+});
+
+beforeEach(async () => {
+  await DatabaseConnection.clear();
+});
+
+afterAll(async () => {
+  await DatabaseConnection.clear();
+  await DatabaseConnection.close();
+});
+
+describe('event CRUD operations', () => {
+  test('events from the past, future, and all time can be pulled', async () => {
+    const conn = await DatabaseConnection.get();
+    const pastEvent = EventFactory.fakePastEvent();
+    const ongoingEvent = EventFactory.fakeOngoingEvent();
+    const futureEvent = EventFactory.fakeFutureEvent();
+    const user = UserFactory.fake();
+
+    await new PortalState()
+      .createEvents([pastEvent, ongoingEvent, futureEvent])
+      .createUsers([user])
+      .write();
+
+    const eventController = ControllerFactory.event(conn);
+
+    const pastEventsResponse = await eventController.getPastEvents({}, user);
+    const futureEventsResponse = await eventController.getFutureEvents({}, user);
+    const allEventsResponse = await eventController.getAllEvents({}, user);
+
+    expect(pastEventsResponse.events).toEqual(
+      expect.arrayContaining([pastEvent.getPublicEvent()]),
+    );
+    expect(futureEventsResponse.events).toEqual(
+      expect.arrayContaining([
+        ongoingEvent.getPublicEvent(),
+        futureEvent.getPublicEvent(),
+      ]),
+    );
+    expect(allEventsResponse.events).toEqual(
+      expect.arrayContaining([
+        pastEvent.getPublicEvent(),
+        ongoingEvent.getPublicEvent(),
+        futureEvent.getPublicEvent(),
+      ]),
+    );
+  });
+
+  test('events can be created with unused attendance codes', async () => {
+    const conn = await DatabaseConnection.get();
+    const [event1] = EventFactory.with({ attendanceCode: 'code' });
+    const [event2] = EventFactory.with({ attendanceCode: 'different-code' });
+    const [admin] = UserFactory.with({ accessType: UserAccessType.ADMIN });
+
+    await new PortalState()
```

Comment:
I think we should prefer controllers wherever we can instead of relying on `PortalState`. Like here, the test's simple enough that we could create via the controller.

Reply:
Hmm, I personally think it's clearer to use `PortalState` for mocking all the data reads/writes that aren't explicitly tested, just so that it is clear what exactly is being tested just by looking at the controller calls that happen within the test. That way we are also reducing the number of dependent controllers that a given test is using so that if a test fails, it is clear that the controllers that are actually being tested are the ones failing (even though like you mentioned, the controller would be used for a simple case).

The one issue I'm seeing with `PortalState` is that it needs to keep being updated to match our API's functionality, but there aren't that many huge changes to the base functionality (like attending events, ordering merch) anyway to warrant a consistent updating of `PortalState`.